Add support for Symfony 7

johanib opened the pull request:
Also update dev & qa tooling.

johanib committed:
Add support for Symfony 7

Also update dev & qa tooling.

johanib commented on composer.json:
```diff
@@ -72,6 +72,13 @@
             ]
-        }
+        },
+        "platform": {
```
weg?

MKodde replied:
I agree with that, I dont think pinning the platform version is good practice in a bundle

MKodde commented on ci/qa/phpcpd:
Maybe not remove the phpcpd, but install the replacement/shim instead?

`https://packagist.org/packages/irstea/phpcpd-shim`

MKodde commented on composer.json:
```diff
@@ -20,46 +20,46 @@
+        "symfony/phpunit-bridge": "^7.3"
     },
     "scripts": {
         "check": [
```
In the SAML library you augmented the scripts with a `check-ci` version. May I suggest to either not do that there, or implement it throughout the bundles/apps? Predictability is quite important imo..

johanib replied:
There was some discussion about rector running in the pipeline. 
It has become a very useful tool, but we don't want to bog down the pipeline with it during periods of low developer engagement. Which is why I opted to run it only in development.

So i'll add it to the other stepup projects as well.

(The same argument could be made for PHPStan, so this might be a double standard...)

MKodde commented on composer.json:
```diff
@@ -20,46 +20,46 @@
-        "symfony/http-kernel": "^5.4|^6.3",
-        "symfony/twig-bridge": "^5.4|^6.3",
-        "symfony/validator": "^5.4|^6.3"
+        "surfnet/stepup-saml-bundle": "dev-feature/maintenance_nov-25",
```
Reminder: Make sure this is removed before tagging a new version of the bundle.

MKodde commented on composer.json:
```diff
@@ -20,46 +20,46 @@
-        "symfony/twig-bridge": "^5.4|^6.3",
-        "symfony/validator": "^5.4|^6.3"
+        "surfnet/stepup-saml-bundle": "dev-feature/maintenance_nov-25",
+        "symfony/config": "^6.3|^7.0",
```
It is nice to allow a broad version range of Symfony versions. 

1. Is this useful/required
2. Do you test if the bundle actually works with SF 6.3, 6.4 & and 7.0? You should be able to make a test matrix that forces different versions from being installed/used. On the other hand, we are probably the exclusive consumer of this bundle. So pinning it to our own desired version might be less of a hassle?

johanib replied:
I will test this in gateway

MKodde replied:
Supporting new and the previous SF version is useful during upgrading of the consumer apps. That way the app can switch to the new bundle version without itself needing to upgrade to the next SF version.

MKodde commented on src/Tests/DateTimeMock.php:
```diff
@@ -0,0 +1,14 @@
+<?php
```
Missing license header. Can be easily prevented by running the license header checker..

johanib replied:
good catch, I added the licence check to the qa pipeline.

MKodde commented on ci/qa/phpstan-baseline.neon:
```diff
@@ -1,301 +1,421 @@
 			path: ../../src/Value/PhoneNumber/CountryCodeListing.php
+
+		-
+			message: '#^Call to function is_string\(\) with string will always evaluate to true\.$#'
+			identifier: function.alreadyNarrowedType
+			count: 1
+			path: ../../src/Value/PhoneNumber/InternationalPhoneNumber.php
+
+		-
+			message: '#^Call to function is_string\(\) with string will always evaluate to true\.$#'
+			identifier: function.alreadyNarrowedType
+			count: 1
+			path: ../../src/Value/PhoneNumber/PhoneNumber.php
+
+		-
+			message: '#^Call to function is_string\(\) with string will always evaluate to true\.$#'
+			identifier: function.alreadyNarrowedType
+			count: 1
+			path: ../../src/Value/SecondFactorType.php
+
+		-
+			message: '#^Call to function is_string\(\) with string will always evaluate to true\.$#'
+			identifier: function.alreadyNarrowedType
+			count: 1
+			path: ../../src/Value/YubikeyOtp.php
```
These might be easy fixes, just removing the incriminating `if` statements in these paths?

johanib replied:
This is a bc break, as the projects using this catch \Surfnet\StepupBundle\Exception\InvalidArgumentException and handle errors accordingly.

Removing the if, and making the string type strict, will cause a different error.

gateway uses this a lot. 

So I was not very confident this will not break stuff. So I did not want to break this for now...

MKodde reviewed (changes requested):
Nice work again! 

Suggestions: 

1. Maybe add an entry in the `CHANGELOG.md`?
2. Run the ci/cd tests on PHP 8.4?

johanib committed:
Rework

- Update changelog and readme
- Add docheader check
- reinstate phpcpd
- update composer

johanib commented:
> Nice work again!
> 
> Suggestions:
> 
> 1. Maybe add an entry in the `CHANGELOG.md`?
> 2. Run the ci/cd tests on PHP 8.4?

:heavy_check_mark: 
Also updated the readme as it was targeted at old Symfony versions.

MKodde reviewed (approved):
Approved, with the promise that you will install the correct SAML bundle version before merging this :)
